radareorg/cutter
# Rewrite raise memory widget

**Detailed description**



Changes in this PR
* Allow seeking without triggering raise memory widget. This was one of the major causes for unexpected changes in focus.
* Raise memory widget mostly moved to main window from core and each widget.
* Last active widget remembered using a pointer instead of memory widget type. Helps when there are multiple widgets with the same type.
* Initial version for "show in" context menu. Currently added only to disassembly and graph menus. Left adding it to other menus for separate PR as it would be good to also refactor some of the common actions for list widgets with addressable objects like functions, strings, imports, search results and others.

I couldn't quite decide if "show in/new disassembly" and others should create a new synchronized or unsynchronized memory widget. Currently Cutter is designed more for synchronized mode. In future the menu could also be reused for lines with jumps (show target). That should probably use unsynchronized. "Show in" and "show target" doesn't necessarily need to behave exactly the same.

**Parts that could be better**

Name for MainWindow and Core `show(address) ` is open for discussions. 

I am not too happy about using event filter for detecting memory widget activation.

**Test plan (required)**

- perform test case described in #1616 
- switching between disassembly and graph using space
  - works in normal case
  - works when either disassembly or graph widget is closed
  - works when disassembly is side by side with graph, test focus by moving up and down
- test rising last active
  - switch to memory widget (repeat with all)
  - switch to overview
  - double click on function
- test ignore empty graph
  - switch to graph
  - switch to symbol list, double click function observe that it switched to graph
  - switch to string list, double click string literal, observe that it switched to memory widget that isn't graph
- test with multiple widgets of the same kind
- disassembly side by side with hex widget, using master it is sometimes impossible to do keyboard navigation in hexwidget because seek moves focus to disassembly
- side by side use case

![showin3](https://user-images.githubusercontent.com/7101031/61330439-3dc52800-a828-11e9-88fc-214b3d0b78aa.png)



**Closing issues**

Closes #1616
